### DISCUSSION TITLE
Using NSError

### DISCUSSION THREAD

**Discussion by dmzkrsk** [progrium/darwinkit#254] [Q&A]
I try to write code, that trashes file and I need to get error details if it fails

```golang
func TrashItem(filePath string) error {
	fileURL := foundation.NewURLFileURLWithPathIsDirectoryRelativeToURL(filePath, false, nil)
	fm := foundation.FileManager_DefaultManager()

	var fe = foundation.NewError()
	ok := fm.TrashItemAtURLResultingItemURLError(fileURL, nil, fe) // also tried &fe

	fmt.Printf("ok=%#v, err=%#v\n", ok, fe)

	if fe.IsNil() {
		fmt.Printf("is nil\n")
	} else {
		fmt.Printf("LocalizedDescription=%#v\n", fe.LocalizedDescription())
        }

	return nil
}
```

But it panics at `fe.LocalizedDescription`

```
SIGSEGV: segmentation violation
PC=0x19093207c m=0 sigcode=2 addr=0x4
signal arrived during cgo execution
```

I use branch 0.5.0

**Comment by programmingkidx** [comment]
Does this work:

ok := fm.TrashItemAtURLResultingItemURLError(fileURL, nil, unsafe.Pointer(&fe))

**Reply by dmzkrsk** [reply]
`cannot use unsafe.Pointer(&fe) (value of type unsafe.Pointer) as foundation.IError value in argument to fm.TrashItemAtURLResultingItemURLError: unsafe.Pointer does not implement foundation.IError (missing method Autorelease)`

**Reply by programmingkidx** [reply]
Do you know the exact git commit of the repo you are using?

**Reply by dmzkrsk** [reply]
https://github.com/progrium/macdriver/tree/v0.5.0-preview

**Reply by programmingkidx** [reply]
I see what is wrong. The function TrashItemAtURLResultingItemURLError should have this signature: func (f_ FileManager) TrashItemAtURLResultingItemURLError(url IURL, outResultingURL unsafe.Pointer, error unsafe.Pointer) bool. It does not. This is a problem I did create an issue for here: https://github.com/progrium/macdriver/issues/255. 

To solve your issue now you would want to run these commands:

1. cd into your macdriver folder.
2. run 'make generate/symbols.zip'
3. run 'go generate ./macos/foundation'
4. Try running your program again.

I created this program to test your function:
```
package main

import (
	"fmt"
        "unsafe"
	"github.com/progrium/macdriver/macos/foundation"
)

func main() {    
    TrashItem("./eraseme.rtf")
}

func TrashItem(filePath string) error {
	fileURL := foundation.NewURLFileURLWithPathIsDirectoryRelativeToURL(filePath, false, nil)
	fm := foundation.FileManager_DefaultManager()

	var fe foundation.Error
        ok := fm.TrashItemAtURLResultingItemURLError(fileURL, nil, unsafe.Pointer(&fe))

	fmt.Printf("ok=%#v, err=%#v\n", ok, &fe)

	if fe.IsNil() {
		fmt.Printf("is nil\n")
	} else {
		fmt.Printf("LocalizedDescription=%#v\n", fe.LocalizedDescription())
        }

	return nil
}
```

I made a file called eraseme.rtf and it was successfully moved into the trash. I only had to change the fe variable to type foundation.Error and wrap it in an unsafe.Pointer(&fe) call in the method call.

Hope this helps.

**Reply by programmingkidx** [reply]
I made a document detailing the problem you experienced. I was wondering if you could review it and let me know if this would have helped you.

https://github.com/progrium/macdriver/wiki/Issues-with-NSError-and-other-arguments-using-pointer-to-pointer-types
